pkg/ext/stores/kubeconfig/store.go

```diff
@@ -0,0 +1,1487 @@
+	}
+
+	host := s.getServerURL()
+	if host != "" {
+		u, err := url.Parse(host)
+		if err == nil {
+			host = u.Host
+		}
+	}
+	if host == "" {
+		host = first(extras[common.ExtraRequestHost])
+		if host == "" {
+			return nil, apierrors.NewInternalError(errors.New("can't determine the server URL"))
+		}
+	}
+
+	var (
+		conditions    []metav1.Condition
+		tokenIDs      []string
+		clusters      []*apiv3.Cluster
+		isAllClusters bool // User requested the kubeconfig for all clusters with "*".
+	)
+
+	localCluster, err := s.clusterCache.Get("local")
+	if err != nil {
+		return nil, apierrors.NewInternalError(fmt.Errorf("error getting local cluster: %w", err))
+	}
+
+	if len(kubeconfig.Spec.Clusters) > 0 {
+		if isAllClusters = kubeconfig.Spec.Clusters[0] == "*"; isAllClusters {
+			// The first id in the spec.clusters "*" means all clusters.
+			clusters, err = s.clusterCache.List(labels.Everything())
+			if err != nil {
+				return nil, apierrors.NewInternalError(fmt.Errorf("error listing clusters: %w", err))
+			}
+		} else {
+			// Individually listed clusters.
+			for _, clusterID := range kubeconfig.Spec.Clusters {
+				if clusterID == "local" { // Shortcut for the local cluster.
+					clusters = append(clusters, localCluster)
+					continue
+				}
+
+				cluster, err := s.clusterCache.Get(clusterID)
+				if err != nil {
+					if apierrors.IsNotFound(err) {
+						return nil, apierrors.NewBadRequest(fmt.Sprintf("cluster %s not found", clusterID))
+					}
+					return nil, apierrors.NewInternalError(fmt.Errorf("error getting cluster %s: %w", clusterID, err))
+				}
+
+				clusters = append(clusters, cluster)
+			}
+		}
+	}
+
+	// The name of the cluster to use as the current context.
+	// Note that the actual context is set later to the display name of the cluster.
+	var currentContext string
+
+	// Check if the user has access to requested clusters before generating tokens.
+	// If the user requested all clusters, figure out which clusters they have access to and adjust the list.
+	for i := 0; i < len(clusters); i++ {
+		decision, _, err := s.authorizer.Authorize(ctx, &authorizer.AttributesRecord{
+			User:            userInfo,
+			Verb:            "get",
+			APIGroup:        mgmt.GroupName,
+			Resource:        apiv3.ClusterResourceName,
+			ResourceRequest: true,
+			Name:            clusters[i].Name,
+		})
+		if err != nil {
+			return nil, apierrors.NewInternalError(fmt.Errorf("error checking if user %s has access to cluster %s: %w", userInfo.GetName(), clusters[i].Name, err))
+		}
+
+		// Check early if the requested current context is valid.
+		if currentContext == "" && kubeconfig.Spec.CurrentContext == clusters[i].Name {
+			currentContext = clusters[i].Name
+		}
+
+		if decision != authorizer.DecisionAllow {
+			if isAllClusters {
+				// Delete the cluster the user doesn't have access to from the list in-place.
+				copy(clusters[i:], clusters[i+1:])
+				clusters[len(clusters)-1] = nil
+				clusters = clusters[:len(clusters)-1]
```

Comment:
I have to admit, I would not have bothered with in-place edit, and rather positively fill a separate result slice with the accessible clusters, with a simple forward iteration of the input slice.